cmd/srcd/cmd/root.go

```diff
@@ -88,37 +88,89 @@ func initConfig() {
 
 var logMsgRegex = regexp.MustCompile(`.*msg="(.+)"`)
 
-func logAfterTimeout(header string) chan struct{} {
-	logs, err := daemon.GetLogs()
-	if err != nil {
-		logrus.Errorf("could not get logs from server container: %v", err)
+func logAfterTimeout(msg string, timeout time.Duration) func() {
+	d := &Defered{
+		Timeout: timeout,
+		Msg:     msg,
 	}
 
-	started := make(chan struct{})
-	go func() {
-		defer logs.Close()
+	return d.Print()
+}
+
+func logAfterTimeoutWithSpinner(msg string, timeout time.Duration, spinnerInterval time.Duration) func() {
+	d := &Defered{
+		Timeout:         timeout,
+		Msg:             msg,
+		Spinner:         true,
+		SpinnerInterval: spinnerInterval,
+	}
+
+	return d.Print()
+}
+
+func logAfterTimeoutWithServerLogs(msg string, timeout time.Duration) func() {
+	logs, err := daemon.GetLogs()
+
+	inputFn := func(stop <-chan bool) <-chan string {
+		ch := make(chan string)
+
+		if err != nil {
+			logrus.Errorf("could not get logs from server container: %v", err)
+			close(ch)
+			return ch
+		}
 
-		select {
-		case <-time.After(3 * time.Second):
-			logrus.Info(header)
+		go func() {
```

Comment:
most probably you can remove one goroutine. You have only 1 blocking operation - reading from scanner. But you create 2 goroutines one inside another.

Reply:
discard this comment